utilities/transactions/write_unprepared_txn.cc

```diff
@@ -519,6 +641,104 @@ void WriteUnpreparedTxn::Clear() {
+  if (save_point_boundaries_ == nullptr) {
+    save_point_boundaries_.reset(new autovector<size_t>());
+  }
+  save_point_boundaries_->push_back(write_batch_.GetDataSize());
+}
+
+Status WriteUnpreparedTxn::RollbackToSavePoint() {
+  Status s = PessimisticTransaction::RollbackToSavePoint();
+  if (!s.IsNotFound()) {
+    return s;
+  }
+
+  if (wup_save_points_ != nullptr && !wup_save_points_->empty()) {
+    return RollbackToSavePointInternal();
+  }
+
+  return Status::NotFound();
+}
+
+Status WriteUnpreparedTxn::RollbackToSavePointInternal() {
+  Status s;
+
+  write_batch_.Clear();
+  WriteBatchInternal::InsertNoop(write_batch_.GetWriteBatch());
+
+  assert(wup_save_points_->size() > 0);
+  WriteUnpreparedTxn::SavePoint &top = wup_save_points_->back();
+
+  assert(top.unprep_seqs_.size() > 0);
+  auto snapshot_seq = WriteUnpreparedTxnReadCallback::CalcMaxVisibleSeq(top.unprep_seqs_, 0);
+
+  // Put TransactionBaseImpl::SavePoint back on the stack so that GetTrackedKeysSinceSavePoint can examine it.
+  assert(save_points_ != nullptr && save_points_->size() > 0);
+  save_points_->emplace(std::move(top.save_point_));
+  const std::unique_ptr<TransactionKeyMap>& tracked_keys = GetTrackedKeysSinceSavePoint();
+  WriteUnpreparedTxnReadCallback callback(wupt_db_, snapshot_seq, kMinUnCommittedSeq, top.unprep_seqs_);
+
+  // TODO(lth): Reduce duplicate code with RollbackInternal logic.
+  ReadOptions roptions;
+  const auto& cf_map = *wupt_db_->GetCFHandleMap();
+  for (const auto& cfkey : *tracked_keys) {
+    const auto cfid = cfkey.first;
+    const auto& keys = cfkey.second;
+    for (const auto& pair : keys) {
+      const auto& key = pair.first;
+      const auto& cf_handle = cf_map.at(cfid);
+      PinnableSlice pinnable_val;
+      bool not_used;
+      s = db_impl_->GetImpl(roptions, cf_handle, key, &pinnable_val, &not_used,
+                            &callback);
+
+      if (s.ok()) {
+        s = write_batch_.Put(cf_handle, key, pinnable_val);
+        assert(s.ok());
+      } else if (s.IsNotFound()) {
+        s = write_batch_.Delete(cf_handle, key);
+        assert(s.ok());
+      } else {
+        return s;
+      }
+    }
+  }
+
+  const bool kPrepared = true;
+  s = FlushWriteBatchToDBInternal(!kPrepared);
+  assert(s.ok());
+  if (!s.ok()) {
+    return s;
+  }
+
+  // We've already replaced TransactionBaseImpl::SavePoint above for GetTrackedKeysSinceSavePoint, but we also need to replace a savepoint for the write_batch, so that we can call PessimisticTransaction::RollbackToSavePoint safely.
+  write_batch_.SetSavePoint();
+  s = PessimisticTransaction::RollbackToSavePoint();
+  assert(s.ok());
+  if (!s.ok()) {
+    return s;
+  }
+
+  wup_save_points_->pop_back();
+  return s;
+}
+
+Status WriteUnpreparedTxn::PopSavePoint() {
+  Status s = PessimisticTransaction::PopSavePoint();
+  if (!s.IsNotFound()) {
+    return s;
+  }
+
+  if (wup_save_points_ != nullptr && !wup_save_points_->empty()) {
```

Comment:
I am confused! Dont we have a WriteUnprepared::SavePoint per TransactionBase::SavePoint?

Reply:
I added more comments in the header to describe the relationship between all the data structures holding savepoint info.